### PR TITLE
Accept xls in clinical parse sch

### PR DESCRIPTION
**Pull request by kairstenfay**
Allow Excel files in the `clinical parse-sch` CLI tool.

### CODE REVIEW

File: lib/seattleflu/id3c/cli/command/clinical.py [hunk 1]
```diff
@@ -115,6 +113,17 @@ def parse_uw(uw_filename, output):
     dump_ndjson(clinical_records)
 
 
+def get_read_function(filename: str) -> Callable:
```
Comment (kairstenfay) [inline lib/seattleflu/id3c/cli/command/clinical.py]: I'm open to better ideas for a name, here

File: lib/seattleflu/id3c/cli/command/clinical.py [hunk 2]
```diff
@@ -199,7 +199,7 @@ def parse_sch(sch_filename, output):
     All clinical records parsed are output to stdout as newline-delimited JSON
     records.  You will likely want to redirect stdout to a file.
     """
-    dtypes = {'census_tract': 'string'}
+    dtypes = {'census_tract': 'str'}
```
Comment (kairstenfay) [inline lib/seattleflu/id3c/cli/command/clinical.py]: When left as `string`, reading a `.xls` file with `pd.read_excel` throws the following error:
```
[2020-02-26 13:58:54] Uncaught exception:
Traceback (most recent call last):
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/bin/id3c", line 11, in <module>
    load_entry_point('id3c', 'console_scripts', 'id3c')()
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/kairsten/seattleflu/id3c-customizations/lib/seattleflu/id3c/cli/command/clinical.py", line 206, in parse_sch
    clinical_records = read(sch_filename, dtype=dtypes)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pandas/io/excel/_base.py", line 334, in read_excel
    **kwds,
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pandas/io/excel/_base.py", line 885, in parse
    **kwds,
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pandas/io/excel/_base.py", line 512, in parse
    output[asheetname] = parser.read(nrows=nrows)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pandas/io/parsers.py", line 1133, in read
    ret = self._engine.read(nrows)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pandas/io/parsers.py", line 2464, in read
    data = self._convert_data(data)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pandas/io/parsers.py", line 2534, in _convert_data
    clean_dtypes,
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pandas/io/parsers.py", line 1727, in _convert_to_ndarrays
    cvals = self._cast_types(cvals, cast_type, c)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pandas/io/parsers.py", line 1825, in _cast_types
    return array_type._from_sequence_of_strings(values, dtype=cast_type)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pandas/core/arrays/string_.py", line 201, in _from_sequence_of_strings
    return cls._from_sequence(strings, dtype=dtype, copy=copy)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pandas/core/arrays/string_.py", line 197, in _from_sequence
    return cls(result)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pandas/core/arrays/string_.py", line 166, in __init__
    self._validate()
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pandas/core/arrays/string_.py", line 171, in _validate
    raise ValueError("StringArray requires a sequence of strings or pandas.NA")
ValueError: StringArray requires a sequence of strings or pandas.NA
```

**Comment by kairstenfay** [comment]
[Read xls or csv in `clinical parse-sch`](https://trello.com/c/3uZSm38H/344-read-xls-or-csv-in-clinical-parse-sch)

**Comment by kairstenfay** [comment]
> Why not use the exisiting `load_file_as_dataframe` function in `id3c.cli.io.pandas`?

I didn't know about that function. Neat!

**Comment by kairstenfay** [comment]
> Why not use the exisiting `load_file_as_dataframe` function in `id3c.cli.io.pandas`?

Hmm when I use it, I get this error:
```
[2020-02-26 14:13:41] Uncaught exception:
Traceback (most recent call last):
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/bin/id3c", line 11, in <module>
    load_entry_point('id3c', 'console_scripts', 'id3c')()
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/kairsten/seattleflu/id3c-customizations/lib/seattleflu/id3c/cli/command/clinical.py", line 207, in parse_sch
    clinical_records = trim_whitespace(clinical_records)
  File "/home/kairsten/seattleflu/id3c-customizations/lib/seattleflu/id3c/cli/command/__init__.py", line 46, in trim_whitespace
    str_columns: List[str] = list(df.select_dtypes(include='string'))
AttributeError: 'dict' object has no attribute 'select_dtypes'
```

**Comment by kairstenfay** [comment]
> > Why not use the exisiting `load_file_as_dataframe` function in `id3c.cli.io.pandas`?
> 
> I didn't know about that function. Neat!

Hmm. `load_file_as_dataframe` is returning my Excel workbook as a dict of type `Dict[str, pd.DataFrame]`...

**Comment by kairstenfay** [comment]
> Hmm. `load_file_as_dataframe` is returning my Excel workbook as a dict of type `Dict[str, pd.DataFrame]`...
> 
> It looks like when we pass `sheet_name=None`, this is Pandas' behavior.

I addressed this bug in https://github.com/seattleflu/id3c/pull/12.

**Comment by kairstenfay** [comment]
On second thought, I'm not so sure `load_file_as_dataframe` is the right function to use. The docstring for our `read_excel` says
> Reads an Excel file while ensuring all values are treated as strings.

I don't know if that's the behavior we want when processing retrospective data.

**Comment by kairstenfay** [comment]
I pushed a WIP commit to show the work I did to try to adopt using `load_file_as_dataframe`.

**Comment by tsibley** [comment]
> On second thought, I'm not so sure `load_file_as_dataframe` is the right function to use. The docstring for our `read_excel` says
> 
> > Reads an Excel file while ensuring all values are treated as strings.
> 
> I don't know if that's the behavior we want when processing retrospective data.

We may not want to treat all columns as strings ultimately, but should be ok to read them that way and convert as necessary.

**Comment by kairstenfay** [comment]
> > On second thought, I'm not so sure `load_file_as_dataframe` is the right function to use. The docstring for our `read_excel` says
> > > Reads an Excel file while ensuring all values are treated as strings.
> > 
> > 
> > I don't know if that's the behavior we want when processing retrospective data.
> 
> We may not want to treat all columns as strings ultimately, but should be ok to read them that way and convert as necessary. (But also I'm ok not doing this.)

Instead of updating our `read_excel` function, I am using `.replace({"": pd.NA})` on the SCH data within `parse-sch`. It has the weird side-effect of converting `<NA>` to empty dicts in JSON.

**Comment by kairstenfay** [comment]
> Instead of updating our `read_excel` function, I am using `.replace({"": pd.NA})` on the SCH data within `parse-sch`. It has the weird side-effect of converting `<NA>` to empty dicts in JSON.

Update: this has nothing to do with `.replace({"": pd.NA})` specifically. So long as I have the `string` data type column of `<NA>` values, the `pd.to_json()` prints these missing values as empty dicts `{}`.

**Comment by kairstenfay** [comment]
> Update: this has nothing to do with `.replace({"": pd.NA})` specifically. So long as I have the `string` data type column of `<NA>` values, the `pd.to_json()` prints these missing values as empty dicts `{}`.

So, what to do here? Stop using `string`, or update our `etl clinical` to account for empty dicts as potential `null` values?

**Comment by kairstenfay** [comment]
I ended up with a simple solution: using `load_file_as_dataframe`, casting empty strings to `None`, and converting age to float.

**Pull request merged**